src/naive_bayes/mod.rs

```diff
@@ -0,0 +1,239 @@
+use crate::error::Failed;
+use crate::linalg::BaseVector;
+use crate::linalg::Matrix;
+use crate::math::num::RealNumber;
+use std::marker::PhantomData;
+
+/// Distribution used in the Naive Bayes classifier.
+pub trait NBDistribution<T: RealNumber, M: Matrix<T>> {
+    /// Prior of class k
+    fn prior(&self, k: T) -> T;
+
+    /// Conditional probability of feature j given class k
+    fn conditional_probability(&self, k: T, j: &M::RowVector) -> T;
+
+    /// Possible classes of the distribution
+    fn classes(&self) -> &Vec<T>;
+}
+
+/// Base struct for the Naive Bayes classifier.
+pub struct BaseNaiveBayes<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> {
+    distribution: D,
+    _phantom_t: PhantomData<T>,
+    _phantom_m: PhantomData<M>,
+}
+
+impl<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> BaseNaiveBayes<T, M, D> {
+    /// Fits NB classifier to a given NBdistribution.
+    /// * `distribution` - NBDistribution of the training data
+    pub fn fit(distribution: D) -> Result<Self, Failed> {
+        Ok(Self {
+            distribution,
+            _phantom_t: PhantomData,
+            _phantom_m: PhantomData,
+        })
+    }
+
+    /// Estimates the class labels for the provided data.
+    /// * `x` - data of shape NxM where N is number of data points to estimate and M is number of features.
+    /// Returns a vector of size N with class estimates.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let y_classes = self.distribution.classes();
+        let (rows, _) = x.shape();
+        let predictions = (0..rows)
+            .map(|row_index| {
+                let row = x.get_row(row_index);
+                let (prediction, _probability) = y_classes
+                    .iter()
+                    .map(|class| {
+                        (
+                            class,
+                            self.distribution.conditional_probability(*class, &row)
+                                * self.distribution.prior(*class),
+                        )
+                    })
+                    .max_by(|(_, p1), (_, p2)| p1.partial_cmp(p2).unwrap())
+                    .unwrap();
+                *prediction
+            })
+            .collect::<Vec<T>>();
+        let mut y_hat = M::RowVector::zeros(rows);
+        for (i, prediction) in predictions.iter().enumerate().take(rows) {
+            y_hat.set(i, *prediction);
+        }
+        Ok(y_hat)
+    }
+}
+
+/// Naive Bayes classifier for categorical features
+pub struct CategoricalNB<T: RealNumber> {
```

Comment:
As I suggested in my earlier comment let's rename this class to CategoricalNBDistribution and make it `pub (crate) ` or private

Reply:
Done, with that change the compiler warns about unused code, but I guess that it is ok for now and it should be solved once we implement the public API

Reply:
Can you also implement public interface?
The goal is to make sure we can fit categorical NB in one line:

```
let nb = CategoricalNB::fit(&x, &y, Default::default()).unwrap();
```

Also, do you think we should move Categorical NB + corresponding distribution into a separate file? Or you'd prefer to keep it in the _mod.rs_ for now?

Reply:
Yes, I can implement that! 

Regarding to the other one, in scikit they have CategoricalNB directly in the naive_bayes module, I think that I can implement them in another file, but I also think that it would be fine to reimport them in the the .mod file, what do you think?

Reply:
I just feel that when we'll add other types of NB the file _mod.rs_ will get too large and tricky to maintain.

Reply:
Implemented in https://github.com/smartcorelib/smartcore/pull/15/commits/7620226e2d27d895749847583b38c52b624ce623